File: FBControlCore/Utility/FBConcurrentCollectionOperations.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBConcurrentCollectionOperations.h"

@interface FBConcurrentCollectionOperations_FilterTerminal : NSObject

+ (instancetype)terminal;

@end

@implementation FBConcurrentCollectionOperations_FilterTerminal

+ (instancetype)terminal
{
  static dispatch_once_t onceToken;
  static FBConcurrentCollectionOperations_FilterTerminal *terminal;
  dispatch_once(&onceToken, ^{
    terminal = [FBConcurrentCollectionOperations_FilterTerminal new];
  });
  return terminal;
}

@end

@implementation FBConcurrentCollectionOperations

#pragma mark Public

+ (NSArray *)generate:(NSUInteger)count withBlock:( id(^)(NSUInteger index) )block
{
  NSMutableArray *array = [NSMutableArray array];
  for (NSUInteger index = 0; index < count; index++) {
    [array addObject:NSNull.null];
  }

  dispatch_apply((size_t) count, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^ (size_t iteration) {
    id object = block(iteration);
    if (object) {
      @synchronized(array) {
        array[iteration] = object;
      }
    }
  });
  return [array copy];
}

+ (NSArray *)map:(NSArray *)array withBlock:( id(^)(id object) )block
{
  return [self
    generate:array.count
    withBlock:^ id (NSUInteger index) {
      return block(array[index]);
    }];
}

+ (NSArray *)filter:(NSArray *)array predicate:(NSPredicate *)predicate
{
  return [self filterMap:array predicate:predicate map:^ id (id object) {
    return object;
  }];
}

+ (NSArray *)mapFilter:(NSArray *)array map:(id (^)(id))block predicate:(NSPredicate *)predicate
{
  NSMutableArray *output = [NSMutableArray array];
  for (NSUInteger index = 0; index < array.count; index++) {
    [output addObject:NSNull.null];
  }

  dispatch_apply(array.count, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^ (size_t iteration) {
    id object = block(array[iteration]);
    BOOL pass = [predicate evaluateWithObject:object];
    if (!pass) {
      object = FBConcurrentCollectionOperations_FilterTerminal.terminal;
    }
    else if (!object) {
      object = NSNull.null;
    }
    @synchronized(output) {
      output[iteration] = object;
    }
  });

  return [output filteredArrayUsingPredicate:self.nonTerminalPredicate];
}

+ (NSArray *)filterMap:(NSArray *)array predicate:(NSPredicate *)predicate map:(id (^)(id))block
{
  return [[self
    map:array
    withBlock:^ id (id object) {
      BOOL pass = [predicate evaluateWithObject:object];
      if (!pass) {
        return FBConcurrentCollectionOperations_FilterTerminal.terminal;
      }
      object = block(object);
      if (!object) {
        return NSNull.null;
      }
      return object;
    }]
    filteredArrayUsingPredicate:self.nonTerminalPredicate];
}

#pragma mark Private

+ (NSPredicate *)nonTerminalPredicate
{
  static dispatch_once_t onceToken;
  static NSPredicate *predicate;
  dispatch_once(&onceToken, ^{
    predicate = [NSPredicate predicateWithBlock:^ BOOL (id evaluatedObject, NSDictionary *_) {
      return evaluatedObject != FBConcurrentCollectionOperations_FilterTerminal.terminal;
    }];
  });
  return predicate;
}

@end

File: FBControlCore/Utility/FBConcurrentCollectionOperations.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import Foundation

private final class FBConcurrentCollectionOperations_FilterTerminal: NSObject, @unchecked Sendable {
  static let terminal = FBConcurrentCollectionOperations_FilterTerminal()
}

private final class UncheckedSendableBox<T>: @unchecked Sendable {
  let value: T
  init(_ value: T) { self.value = value }
}

@objc(FBConcurrentCollectionOperations)
public final class FBConcurrentCollectionOperations: NSObject {

  @objc(generate:withBlock:)
  public class func generate(_ count: UInt, withBlock block: @Sendable @escaping (UInt) -> Any) -> [Any] {
    let array = NSMutableArray(capacity: Int(count))
    for _ in 0..<count {
      array.add(NSNull())
    }

    let sendableArray = UncheckedSendableBox(array)
    DispatchQueue.concurrentPerform(iterations: Int(count)) { iteration in
      let object = block(UInt(iteration))
      objc_sync_enter(sendableArray.value)
      sendableArray.value[iteration] = object
      objc_sync_exit(sendableArray.value)
    }
    return Array(array)
  }

  @objc(map:withBlock:)
  public class func map(_ array: [Any], withBlock block: @Sendable @escaping (Any) -> Any) -> [Any] {
    let sendableArray = UncheckedSendableBox(array)
    return generate(UInt(array.count)) { index in
      return block(sendableArray.value[Int(index)])
    }
  }

  @objc(filter:predicate:)
  public class func filter(_ array: [Any], predicate: NSPredicate) -> [Any] {
    return filterMap(array, predicate: predicate) { $0 }
  }

  @objc(mapFilter:map:predicate:)
  public class func mapFilter(_ array: [Any], map block: @Sendable @escaping (Any) -> Any, predicate: NSPredicate) -> [Any] {
    let output = NSMutableArray(capacity: array.count)
    for _ in 0..<array.count {
      output.add(NSNull())
    }

    let sendableOutput = UncheckedSendableBox(output)
    let sendablePredicate = UncheckedSendableBox(predicate)
    let sendableArray = UncheckedSendableBox(array)
    DispatchQueue.concurrentPerform(iterations: array.count) { iteration in
      var object: Any = block(sendableArray.value[iteration])
      let pass = sendablePredicate.value.evaluate(with: object)
      if !pass {
        object = FBConcurrentCollectionOperations_FilterTerminal.terminal
      }
      objc_sync_enter(sendableOutput.value)
      sendableOutput.value[iteration] = object
      objc_sync_exit(sendableOutput.value)
    }

    return Array(output).filter { !($0 is FBConcurrentCollectionOperations_FilterTerminal) }
  }

  @objc(filterMap:predicate:map:)
  public class func filterMap(_ array: [Any], predicate: NSPredicate, map block: @Sendable @escaping (Any) -> Any) -> [Any] {
    let sendablePredicate = UncheckedSendableBox(predicate)
    let mapped = self.map(array) { object -> Any in
      if !sendablePredicate.value.evaluate(with: object) {
        return FBConcurrentCollectionOperations_FilterTerminal.terminal
      }
      return block(object)
    }
    return mapped.filter { !($0 is FBConcurrentCollectionOperations_FilterTerminal) }
  }
}
